src/utils/math.ts

```diff
@@ -20,7 +23,49 @@ const clamp = (num: number, min: number, max: number) => Math.min(Math.max(num,
 const getChunkNum =
     (itemNum: number, chunkSize: number) => Math.max(1, Math.ceil(itemNum / chunkSize));
 
+/**
+ * Finds the index of the last element in a sorted collection that is less than or equal to
+ * `upperboundValue`. If all elements in the collection are greater than `upperboundValue`, return
+ * the first index of the collection (i.e., `0`).
+ *
+ * @param get Function provided to access elements in the collection by index.
+ * @param lowIdx
+ * @param highIdx
+ * @param upperboundValue
+ * @return The index of the last element less than or equal to `upperboundValue`,`0` if all elements
+ * are greater, or `null` if the collection is empty or the indices are invalid.
+ */
+const upperBoundBinarySearch = <T>(
+    get: (index: number) => T,
+    lowIdx: number,
+    highIdx: number,
+    upperboundValue: T,
+): Nullable<number> => {
+    if (highIdx < lowIdx || "undefined" === typeof (get(highIdx))) {
```

Comment:
Based on this condition, I guess `highIdx` is inclusive. This is the kind of detail that should be included in the docstring; otherwise users have to read the code to figure it out, negating the point of the docstring.